File: fixtures/r_demo/analysis/qa.R
# Stage 2b — a SECOND, independent consumer of processed_iris.csv.
# report.R and qa.R never call each other. They are coupled ONLY through the
# columns of the data product — exactly the connection the web view must surface.

suppressPackageStartupMessages({
  library(dplyr)
  library(readr)
})

PROCESSED <- "processed_iris.csv"

load_processed <- function(path = PROCESSED) {
  read_csv(path, show_col_types = FALSE)
}

#' Also reads petal_ratio — the shared column that couples qa to report.
check_ratios <- function(df = NULL) {
  if (is.null(df)) df <- load_processed()
  bad <- df %>% filter(petal_ratio <= 0)
  list(rows = nrow(df), bad_petal_ratio = nrow(bad))
}

check_species <- function(df = NULL) {
  if (is.null(df)) df <- load_processed()
  df %>% count(species)
}

File: fixtures/python_demo/analysis/qa.py
"""Stage 2b — a SECOND, independent consumer of processed_iris.csv.

report.py and qa.py never call each other. They are coupled ONLY through the
columns of the data product — exactly the connection the web view must surface.
"""
from __future__ import annotations

from pathlib import Path

import pandas as pd

PROCESSED = Path(__file__).resolve().parents[1] / "processed_iris.csv"


def load_processed(path=PROCESSED):
    return pd.read_csv(path)


def check_ratios(df=None):
    """Also reads petal_ratio — the shared column that couples qa to report."""
    df = load_processed() if df is None else df
    bad = df[df["petal_ratio"] <= 0]
    return {"rows": len(df), "bad_petal_ratio": len(bad)}


def check_species(df=None):
    df = load_processed() if df is None else df
    return df["species"].value_counts().to_dict()
